samples-js/functions/weatherMonitor.js

```diff
@@ -0,0 +1,129 @@
+const { output } = require("@azure/functions");
+const df = require("durable-functions");
+const { DateTime } = require("luxon");
+const axios = require("axios").default;
+
+const clearWeatherConditions = ["clear sky", "few clouds", "scattered clouds", "broken clouds"];
+
+const getIsClearActivity = "getIsClear";
+const sendGoodWeatherAlertActivity = "sendGoodWeatherAlert";
+
+if (
+    process.env.OpenWeatherMapApiKey &&
+    process.env.TwilioPhoneNumber &&
+    process.env.TwilioAccountSid &&
+    process.env.TwilioAuthToken
+) {
+    df.app.orchestration("weatherMonitor", function* (context) {
+        const input = context.df.getInput();
+        context.log(`Received monitor request. Location: ${input.location}. Phone: ${input.phone}`);
+        verifyRequest(input);
+
+        const endTime = DateTime.fromJSDate(context.df.currentUtcDateTime).plus({ hours: 6 });
+
+        const locationString = `${input.location.city}${
+            input.location.state ? `, ${input.location.state}` : ""
+        }, ${input.location.country}`;
+
+        context.log(`Instantiating monitor for ${locationString}. Expires: ${endTime.toString()}.`);
+
+        while (DateTime.fromJSDate(context.df.currentUtcDateTime) < endTime) {
+            // Check the weather
+            context.log(
+                `Checking current weather conditions for ${locationString} at ${context.df.currentUtcDateTime}.`
+            );
+            const isClear = yield context.df.callActivity(getIsClearActivity, input.location);
+
+            if (isClear) {
+                // It's not raining! Or snowing. Or misting. Tell our user to take advantage of it.
+                context.log(
+                    `Detected clear weather for ${locationString}. Notifying ${input.phone}.`
+                );
+
+                yield context.df.callActivity(sendGoodWeatherAlertActivity, input.phone);
+                break;
+            } else {
+                // Wait for the next checkpoint
+                const nextCheckpoint = DateTime.fromJSDate(context.df.currentUtcDateTime).plus({
+                    seconds: 30,
+                });
+
+                context.log(`Next check for ${locationString} at ${nextCheckpoint.toString()}`);
+
+                yield context.df.createTimer(nextCheckpoint.toJSDate()); // accomodate cancellation tokens
+            }
+        }
+
+        context.log("Monitor expiring.");
+    });
```

Comment:
Can we sprinkle more comments here to help guide the user through this logic? Given the "height" of this method, I find it a bit difficult to read. I think comments describing logical blocks would help.

Reply:
I added some more comments, let me know if those help! Also, we don't _have_ to have all the functions in one file just because now we can. Let me know if it might make more sense/make it more readable to split each of the functions in this example into its own file!

Reply:
I think it's fine as is. This is certainly a visually noisy sample, so we may want to rethink eventually, but it's fine for now